src/RandomInitialization.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Bdelespierre\Kmeans;
+
+use Bdelespierre\Kmeans\Cluster;
+use Bdelespierre\Kmeans\Interfaces\ClusterCollectionInterface;
+use Bdelespierre\Kmeans\Interfaces\InitializationSchemeInterface;
+use Bdelespierre\Kmeans\Interfaces\PointCollectionInterface;
+use Bdelespierre\Kmeans\Interfaces\PointInterface;
+use Bdelespierre\Kmeans\Point;
+
+class RandomInitialization implements InitializationSchemeInterface
+{
+    /**
+     * @throws \InvalidArgumentException when $nbClusters is lesser than 1
+     */
+    public function initializeClusters(PointCollectionInterface $points, int $nbClusters): ClusterCollectionInterface
+    {
+        // validate cluster count
+        if ($nbClusters < 1) {
+            throw new \InvalidArgumentException("Invalid cluster count: {$nbClusters}");
+        }
+
+        $clusters = new ClusterCollection($points->getSpace());
+
+        // get the space boundaries to avoid placing clusters centroid too far from points
+        list($min, $max) = $this->getBoundaries($points);
+
+        // initialize N clusters with a random point within space boundaries
+        for ($n = 0; $n < $nbClusters; $n++) {
+            // assign all points to the first cluster only
+            $clusters->add(new Cluster($this->getRandomPoint($min, $max), $n == 0 ? $points : null));
+        }
+
+        return $clusters;
+    }
+
+    /**
+     * @return array{PointInterface, PointInterface}
+     * @throws \InvalidArgumentException when $points is empty
+     */
+    protected function getBoundaries(PointCollectionInterface $points): array
+    {
+        if (!count($points)) {
+            throw new \LogicException("Cannot determine boundaries: points collection is empty");
+        }
+
+        /** @var array<array<float>> */
+        $coordinates = array_map(fn ($point) => $point->getCoordinates(), iterator_to_array($points));
+
+        return array_map(
+            fn ($coordinates) => new Point($points->getSpace(), $coordinates),
+            find_boundaries($coordinates)
+        );
+    }
+
+    protected function getRandomPoint(PointInterface $min, PointInterface $max): PointInterface
```

Comment:
Isn't it okay to be private?

Reply:
Same thing here. I do not want to lock future implementations and force the redefinition of `initializeCluster`. We don't know how the community is going to use those classes.